// @sanity/vision/lib/components/QueryErrorDetails.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function QueryErrorDetails(props) {
  const details = props.error.details;
  if (!details || !details.line) {
    return null;
  }

  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'pre',
      null,
      _react2.default.createElement(
        'code',
        null,
        `${details.line}\n${dashLine(details.column)}`
      )
    ),
    _react2.default.createElement(
      'pre',
      null,
      _react2.default.createElement(
        'code',
        null,
        `Line:   ${details.lineNumber}\nColumn: ${details.column}`
      )
    )
  );
}

function dashLine(ln) {
  const line = new Array(ln + 1).join('-');
  return `${line}^`;
}

QueryErrorDetails.propTypes = {
  error: _propTypes2.default.shape({
    details: _propTypes2.default.shape({
      line: _propTypes2.default.string,
      lineNumber: _propTypes2.default.number,
      column: _propTypes2.default.column
    })
  })
};

exports.default = QueryErrorDetails;
