// @sanity/core/lib/actions/cors/addCorsOrigin.js
'use strict';

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const url = require('url');

module.exports = (() => {
  var _ref = _asyncToGenerator(function* (givenOrigin, context) {
    const apiClient = context.apiClient,
          prompt = context.prompt;

    const origin = yield givenOrigin ? filterAndValidateOrigin(givenOrigin) : promptForOrigin(prompt);

    const client = apiClient({
      requireUser: true,
      requireProject: true
    });

    return client.request({
      method: 'POST',
      url: '/cors',
      body: { origin },
      maxRedirects: 0
    });
  });

  function addCorsOrigin(_x, _x2) {
    return _ref.apply(this, arguments);
  }

  return addCorsOrigin;
})();

function promptForOrigin(prompt) {
  return prompt.single({
    type: 'input',
    message: 'Origin (including protocol):',
    filter: filterOrigin,
    validate: validateOrigin
  });
}

function filterOrigin(origin) {
  if (origin === '*') {
    return '*';
  }

  try {
    const parsed = url.parse(origin);
    if (!/^https?:$/.test(parsed.protocol || '')) {
      return null;
    }

    const host = parsed.host.replace(/:(80|443)$/, '');
    return `${parsed.protocol}//${host}`;
  } catch (err) {
    return null;
  }
}

function validateOrigin(origin) {
  if (origin === '*') {
    return true;
  }

  try {
    url.parse(origin || 0); // Use 0 to trigger error for unset values
    return true;
  } catch (err) {
    // Fall-through to error
  }

  return 'Invalid origin, must include protocol (http://some.host)';
}

function filterAndValidateOrigin(givenOrigin) {
  const origin = filterOrigin(givenOrigin);
  const result = validateOrigin(origin);
  if (result !== true) {
    throw new Error(result);
  }

  return origin;
}
