// @sanity/vision/lib/SanityVision.js
'use strict';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _client = require('part:@sanity/base/client');

var _client2 = _interopRequireDefault(_client);

var _default = require('part:@sanity/components/buttons/default');

var _default2 = _interopRequireDefault(_default);

var _schema = require('part:@sanity/base/schema?');

var _schema2 = _interopRequireDefault(_schema);

var _Select = require('./sanity/Select');

var _Select2 = _interopRequireDefault(_Select);

var _Vision = require('./Vision');

var _Vision2 = _interopRequireDefault(_Vision);

var _visionGui = require('./css/visionGui.css');

var _visionGui2 = _interopRequireDefault(_visionGui);

var _jsonInspector = require('./css/jsonInspector.css');

var _jsonInspector2 = _interopRequireDefault(_jsonInspector);

var _jsonDump = require('./css/jsonDump.css');

var _jsonDump2 = _interopRequireDefault(_jsonDump);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const components = {
  Button: _default2.default,
  Select: _Select2.default
};

const styles = {
  jsonDump: _jsonDump2.default,
  visionGui: _visionGui2.default,
  jsonInspector: _jsonInspector2.default
};

const client = _client2.default.clone();

// Used in Sanity project
function SanityVision() {
  return _react2.default.createElement(_Vision2.default, { styles: styles, components: components, client: client, schema: _schema2.default });
}

module.exports = SanityVision;
