// @sanity/vision/lib/containers/VisionContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _DelayedSpinner = require('../components/DelayedSpinner');

var _DelayedSpinner2 = _interopRequireDefault(_DelayedSpinner);

var _ErrorDialog = require('../components/ErrorDialog');

var _ErrorDialog2 = _interopRequireDefault(_ErrorDialog);

var _VisionGui = require('../components/VisionGui');

var _VisionGui2 = _interopRequireDefault(_VisionGui);

var _LoadingContainer = require('./LoadingContainer');

var _LoadingContainer2 = _interopRequireDefault(_LoadingContainer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Loads the most basic data from a Sanity project
class VisionContainer extends _LoadingContainer2.default {
  getSubscriptions() {
    return {
      datasets: { uri: '/datasets' }
    };
  }

  render() {
    if (this.state.error) {
      return _react2.default.createElement(_ErrorDialog2.default, {
        heading: 'An error occured while loading project data',
        error: this.state.error
      });
    }

    if (!this.hasAllData()) {
      return _react2.default.createElement(_DelayedSpinner2.default, null);
    }

    return _react2.default.createElement(_VisionGui2.default, _extends({}, this.state, this.props));
  }
}

exports.default = VisionContainer;
